### ISSUE TITLE
Automatic Service Principal creation in the Azure deployment

### ISSUE THREAD

**Issue by pierocor** [informatics-lab/aml-jupyterhub#43]
The missing step to achieve a one-click deployment of JupyterHub on Azure is the automatic creation of a Service Principal to register our App.
In particular, the above consists of 3 steps:
1. create a SP with the Contributor role (on the Resource Group);
2. delegate permission of 'sign in and read user profile';
3. set the reply URL for authentication.

The [`setenv.sh` script](https://github.com/informatics-lab/aml-jupyterhub/blob/feature/deploy-to-azure-sp-creation/setenv.sh) carries out the above via Azure CLI:
```
az ad sp create-for-rbac --name  $SERVICE_PRINCIPAL_NAME  --role contributor --scopes "/subscriptions/${SUBSCRIPTION_ID}"
az ad app permission add --id=$AAD_CLIENT_ID --api="00000002-0000-0000-c000-000000000000"  --api-permissions="311a71cc-e848-46a1-bdf8-97ff7156d8e6=Scope"
az ad app permission grant --id=$AAD_CLIENT_ID --api="00000002-0000-0000-c000-000000000000"
az ad app update --id=$AAD_CLIENT_ID --reply-urls="https://$HOST/hub/oauth_callback"
```

Apparently ([GH](https://github.com/Azure/azure-quickstart-templates/issues/4981)), it is not possible to create a SP with an ARM template. The only way I found ([AZ docs](https://docs.microsoft.com/en-us/azure/role-based-access-control/role-assignments-template#new-service-principal)) is to define managed identity but this does not seem to be what we need (step 1 is clear, how to achieve 2 and 3?).

Another solution is to use Azure CLI inside the VM to run the above commands. Once defined a managed identity with Owner role into the template this can be assigned to the VM ([AZ docs](https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/qs-configure-template-windows-vm)) and used to log in to Azure via CLI. Unfortunately, this is not enough as the first command returns an error
```
Insufficient privileges to complete the operation.
```
[From the docs](https://docs.microsoft.com/en-us/azure/role-based-access-control/troubleshooting#problems-with-azure-role-assignments) there are two ways to resolve this error:
- to assign the Directory Readers role to the managed identity (but this requires the user to have Administrator role on the RG);
- to replace the --assignee parameter by --assignee-object-id (I am not sure what this means in the above case).

**Comment by tam203** [comment]
@pierocor - thanks for this. On "create a SP with the Contributor role (on the Resource Group);" do you mean subscription rather than resource group?

**Comment by pierocor** [comment]
You are right. In our current tests, the RG scope is sufficient but the original design of the App would require the SP to have contributor role at subscription level to create CIs on behalf of a user on different "projects" (i.e. resource groups).

**Comment by tam203** [comment]
Hi Theo

 

Just to confirm that I think this is not possible.

As far as I understand it, ARM templates are for managing resources within a subscription, whereas the SP sits under the tenant, and so using a mix of powershell/devops/cli + Arm template is the way to go!

 

You’ve probably already found this page [Is there a way to use ARM Template to create an Azure Service Principal?](https://stackoverflow.com/questions/63565067/is-there-a-way-to-use-arm-template-to-create-an-azure-service-principal) - Stack Overflow

but it has a few useful pointers.

 

Edit – actually, one of my colleagues has just come back to me on this - there may be a way to do it – have you seen this new feature?

[Use deployment scripts in templates - Azure Resource Manager | Microsoft Docs](https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/deployment-script-template?tabs=CLI)

It says that you can use it to create objects in Azure AD (preview)

 

Let me know if that gets you somewhere with it!

 

Kind regards

 

Linda

**Comment by pierocor** [comment]
I tried using the deployment script (at resource group level) but with no success. The same error as above appears:
```
Deployment failed. Correlation ID: 0eb66241-6bfe-4f4d-a68b-4863a81f6d01. {
  "status": "failed",
  "error": {
    "code": "ResourceDeploymentFailure",
    "message": "The resource operation completed with terminal provisioning state 'failed'.",
    "details": [
      {
        "code": "DeploymentScriptError",
        "message": "The provided script failed with the following error:\r\nERROR: Insufficient privileges to complete the operation.. Please refer to https://aka.ms/DeploymentScriptsTroubleshoot for more deployment script information."
      }
    ]
  }
}
```
A possible way forward is to assign the role to an existing managed identity via a deployment at subscription level (as described [here](https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/deploy-to-subscription?tabs=azure-cli#access-control)).
